Code/GraphMol/CIPLabeler/CIPMol.cpp

```diff
@@ -0,0 +1,144 @@
+//
+//
+//  Copyright (C) 2020 Schrödinger, LLC
+//
+//   @@ All Rights Reserved @@
+//  This file is part of the RDKit.
+//  The contents are covered by the terms of the BSD license
+//  which is included in the file license.txt, found at the root
+//  of the RDKit source tree.
+//
+
+#include <GraphMol/MolOps.h>
+
+#include "CIPMol.h"
+
+namespace RDKit {
+namespace CIPLabeler {
+
+CIPMol::CIPMol(ROMol *mol) : mol{mol} {}
+
+Fraction CIPMol::getFractionalAtomicNum(Atom *atom) const {
+  PRECONDITION(atom, "bad atom")
+  if (atomnums.empty())
+    const_cast<CIPMol *>(this)->atomnums = calcFracAtomNums(this);
+  return atomnums[atom->getIdx()];
+}
+
+unsigned CIPMol::getNumAtoms() const { return mol->getNumAtoms(); }
+
+unsigned CIPMol::getNumBonds() const { return mol->getNumBonds(); };
+
+Atom *CIPMol::getAtom(int idx) const { return mol->getAtomWithIdx(idx); };
+
+CXXAtomIterator<MolGraph, Atom *> CIPMol::atoms() const { return mol->atoms(); }
+
+Bond *CIPMol::getBond(int idx) const { return mol->getBondWithIdx(idx); };
+
+CIPMolIterator<Bond *, ROMol::OEDGE_ITER> CIPMol::getBonds(Atom *atom) const {
+  return {mol, mol->getAtomBonds(atom)};
+}
+
+CIPMolIterator<Atom *, ROMol::ADJ_ITER> CIPMol::getNeighbors(Atom *atom) const {
+  return {mol, mol->getAtomNeighbors(atom)};
+}
+
+bool CIPMol::isInRing(Bond *bond) const {
+  const auto rings = mol->getRingInfo();
+
+  if (!rings->isInitialized()) {
+    MolOps::findSSSR(*mol);
+  }
+
+  return rings->numBondRings(bond->getIdx()) != 0u;
+};
+
+int CIPMol::getBondOrder(Bond *bond) const {
+  if (kekulized_mol == nullptr) {
+    auto tmp = new RWMol(*mol);
+    MolOps::Kekulize(*tmp);
+    const_cast<CIPMol *>(this)->kekulized_mol.reset(tmp);
+  }
+
+  const auto kekulized_bond = kekulized_mol->getBondWithIdx(bond->getIdx());
+  switch (kekulized_bond->getBondType()) {
+  case Bond::ZERO:
+  case Bond::HYDROGEN:
+  case Bond::DATIVE:
```

Comment:
RIght. I missed those... I never saw any molecules using them (although, in fact, I don't remember ever seeing a dative bond in any molecules I loaded into rdkit)

Reply:
They don't show up in a lot of data sets we use, but they definitely do occur.
I think setting DATIVEL and DATIVER to zero, as Dan suggests, makes sense, but DATIVE is not automatically zero.
The logic technically should be that a DATIVE bond is effectively order 0 for its begin atom, but order 1 for its end atom (i.e. it is asymmetric and only impacts the end atom). 
Here's a little demo of that, notice the explicit valences on the atoms:
```
In [3]: m = Chem.MolFromSmiles('[Fe]<-[NH3]')

In [4]: m.Debug()
Atoms:
        0 26 Fe chg: 0  deg: 1 exp: 1 imp: 0 hyb: 4 arom?: 0 chi: 0
        1 7 N chg: 0  deg: 1 exp: 3 imp: 0 hyb: 5 arom?: 0 chi: 0
Bonds:
        0 1->0 order: 17 conj?: 0 aromatic?: 0
```

Reply:
I'm not 100% percent sure, but I think it might not be relevant, since bond orders are only used to build the graph, and to calculate atomic numbers mancude rings.

I think for now I will just will add a comment mentioning that the bond might not have order 0 for the end atom.